File: app/src/main/java/com/games/memorygame/ui/widget/BoardView.java
package com.games.memorygame.ui.widget;

import android.content.Context;
import android.os.Handler;
import android.util.AttributeSet;
import android.support.v7.widget.GridLayout;

import com.games.memorygame.R;
import com.games.memorygame.model.Photo;
import com.games.memorygame.ui.event.CompletionEvent;
import com.games.memorygame.ui.event.CardPairSelectionEvent;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.List;


public class BoardView extends GridLayout {

    List<BoardCardView> flippedCards = new ArrayList<>();
    private int cardsToFlip = 0;

    public BoardView(Context context) {
        super(context);
    }

    public BoardView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public BoardView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void setUp(List<Photo> photos, int rows, int columns) {
        cardsToFlip = photos.size();
        setOrientation(HORIZONTAL);
        setColumnCount(columns);
        setRowCount(rows);
        for (int row = 0; row < rows; row++) {
            for (int column = 0; column < columns; column++) {
                addView(buildView(photos.get(row * (rows - 1) + column), row, column));
            }
        }
    }

    private boolean canFlip(BoardCardView view) {
        return flippedCards.size() < 2 && view.isFlippedDown();
    }

    private BoardCardView buildView(Photo photo, int row, int column) {
        BoardCardView view = (BoardCardView) inflate(getContext(), R.layout.item_list_card, null);
        int margin = 10;
        GridLayout.LayoutParams params =
                new LayoutParams(GridLayout.spec(row, 1f), GridLayout.spec(column, 1f));
        params.height = 0;
        params.width = 0;
        params.setMargins(margin, margin, margin, margin);
        view.setLayoutParams(params);
        view.setUp(photo);
        view.setOnClickListener(v -> {
            if (!canFlip(view)) {
                return;
            }
            flippedCards.add(view);
            view.flipUp();
            checkMatches();
        });
        return view;
    }

    private void checkMatches() {
        if (flippedCards.size() != 2) {
            return;
        }
        BoardCardView first = flippedCards.get(0);
        BoardCardView second = flippedCards.get(1);
        final Handler handler = new Handler();
        handler.postDelayed(() -> {
            boolean matches = first.match(second);
            EventBus.getDefault().post(new CardPairSelectionEvent(matches));
            if (matches) {
                onMatch();
                return;
            }
            first.flipDown();
            second.flipDown();
            flippedCards.clear();
        }, 1500);
    }

    private void onMatch() {
        cardsToFlip -= 2;
        flippedCards.get(0).setVisibility(INVISIBLE);
        flippedCards.get(1).setVisibility(INVISIBLE);
        if (cardsToFlip == 0) {
            EventBus.getDefault().post(new CompletionEvent());
        }
        flippedCards.clear();
    }
}

File: app/src/main/java/com/games/memorygame/ui/widget/BoardView.kt
package com.games.memorygame.ui.widget

import android.content.Context
import android.os.Handler
import android.util.AttributeSet
import android.support.v7.widget.GridLayout
import android.view.View

import com.games.memorygame.R
import com.games.memorygame.model.Photo
import com.games.memorygame.ui.event.CompletionEvent
import com.games.memorygame.ui.event.CardPairSelectionEvent

import org.greenrobot.eventbus.EventBus

import java.util.ArrayList


class BoardView : GridLayout {

    private var flippedCards: MutableList<BoardCardView> = ArrayList()
    private var cardsToFlip = 0

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?) : super(context)


    fun setUp(photos: List<Photo>, rows: Int, columns: Int) {
        cardsToFlip = photos.size
        orientation = GridLayout.HORIZONTAL
        columnCount = columns
        rowCount = rows
        (0..rows).forEach { r ->
            (0..columns).forEach { c -> addView(buildView(photos[r * (rows - 1) + c], r, c)) }
        }
    }

    private fun canFlip(view: BoardCardView): Boolean {
        return flippedCards.size < 2 && view.isFlippedDown
    }

    private fun buildView(photo: Photo, row: Int, column: Int): BoardCardView {
        val view = View.inflate(context, R.layout.item_list_card, null) as BoardCardView
        val margin = 10
        val params = GridLayout.LayoutParams(GridLayout.spec(row, 1f), GridLayout.spec(column, 1f))
        params.height = 0
        params.width = 0
        params.setMargins(margin, margin, margin, margin)
        view.layoutParams = params
        view.setUp(photo)
        view.setOnClickListener { _ ->
            if (!canFlip(view)) {
                return@setOnClickListener
            }
            flippedCards.add(view)
            view.flipUp()
            checkMatches()
        }
        return view
    }

    private fun checkMatches() {
        if (flippedCards.size != 2) {
            return
        }
        val first = flippedCards[0]
        val second = flippedCards[1]
        val handler = Handler()
        handler.postDelayed({
            val matches = first.match(second)
            EventBus.getDefault().post(CardPairSelectionEvent(matches))
            if (matches) {
                onMatch()
                return@postDelayed
            }
            first.flipDown()
            second.flipDown()
            flippedCards.clear()
        }, 1500)
    }

    private fun onMatch() {
        cardsToFlip -= 2
        flippedCards[0].visibility = View.INVISIBLE
        flippedCards[1].visibility = View.INVISIBLE
        if (cardsToFlip == 0) {
            EventBus.getDefault().post(CompletionEvent())
        }
        flippedCards.clear()
    }
}
